README.md
# Stop "Fiddling" With It Twitter - sfwiT

This repository contains a small Twitter client.  It uses python3 and Node.js.  Python serves as the interface between the Twitter server and my node.js web server (which uses express) that displays the information. My Twitter interface uses python-twitter which is documented at https://python-twitter.readthedocs.io/en/latest/

I run this client on one of my Raspberry PIs on my home network.  From my phone or any browsing device, I browse to that computer's port 2999 to see the latest Tweets from those I follow.  The main page displays a list of those I follow.  When I select from the list, their latest tweets are displayed.  This particular client does not tweet, I very rarely tweet since I'm mostly interested in "news".

I have a whole series of applications that I run as servers on my home network.  This application series, SFWI (stop 'fiddling' with it), provides simple applications that do only what I want and don't leak private information.  I use this application to avoid advertising, Twitter produced trends, and "customized" home timelines.  Twitter insists that it knows what I want to see.  It doesn't.  I don't want to see the same stuff over and over, I don't want to be only associating with people of "like mind".  This client is my first attempt at providing the interface I want.

# Setup

I'm running this on a Raspberry PI under Raspbian (buster).  You don't have to have a Raspberry PI or Raspbian.  This should run fine on most Debian Linux systems.

You will need git, python3, pip3, Node.js, and npm.

From a bash shell:

1) git clone https://github.com/mbroihier/sfwiT
2) cd sfwiT
3) ./configure (if you supply a port number, this alternative port number will be used for the web server)
4) Add the necessary Twitter tokens to TwitterToken.py with your favorite editor
5) Start the client using the wrapper script
   - ./sfwiTWrapper
6) Use a browser to see a list of your "friends" and their timelines
   - http://myComputer:2999 (in your browser)


debian/changelog
sfwit (0.1-3) stable; urgency=low

  [ Mark Broihier ]
  * Twitter Client

 -- Mark Broihier <mbroihier@yahoo.com>  Tue, 08 Sep 2020 17:10:00 +0000


SFWIT.py
'''
Twitter interface for sfwiT
'''
import json
import twitter
import TwitterTokens as T

class SFWIT ():
        '''
        Stop Fiddling With It Twitter class
        '''
        def __init__(self):
            '''
            SFWIT constructor
            '''
            self.api = twitter.Api(consumer_key=T.CONSUMER_KEY, consumer_secret=T.CONSUMER_SECRET, access_token_key=T.ACCESS_TOKEN_KEY, access_token_secret=T.ACCESS_TOKEN_SECRET, tweet_mode='extended')
            self.following = []
        def getEmbeddedStatus(self, statusID):
            '''
            Get a status record in an embedded format
            '''
            try:
                status = self.api.GetStatusOembed(status_id=int(statusID))
            except twitter.error.TwitterError as err:
                print("Execption: ", err, err.message, err.args, statusID, int(statusID))
                #print(dir(err))
                status = err.message
            return json.dumps(status)
        def getUserTimeline(self, screenName):
            '''
            Get the timeline text of someone that is being followed
            '''
            timeLine =  self.api.GetUserTimeline(screen_name=screenName)
            textList = []
            for entry in timeLine:
                if 'retweeted_status' in entry.AsDict():
                    textList.append((entry.created_at, entry.full_text, entry.AsDict()['retweeted_status']['full_text'], str(entry.id)))
                else:
                    textList.append((entry.created_at, entry.full_text, str(entry.id)))
            return json.dumps(textList)
        def printUserTimeline(self, timelineList):
            '''
            Print a list generated from getUserTimeline - assumes JSON format
            '''
            timelineList = json.loads(timelineList)
            for entry in timelineList:
                if len(entry) == 3:
                    print(entry[0])
                    print(entry[1])
                    print(entry[2])
                else:
                    print(entry[0])
                    print(entry[1])
                    print(entry[2])
                    print(entry[3])
        def getFollowing(self):
            '''
            Get the list of screen names being followed
            '''
            self.following = []
            for friend in self.api.GetFriends():
                self.following.append(friend.AsDict()['screen_name'])
            return json.dumps(self.following)


twitterClient.py
'''
Twitter Client
'''
import asyncio
import getopt
import socket
import sys
import websockets

BUFFERSIZE = 1024

@asyncio.coroutine
def web_socket_client(message):
    '''
    Web sock client to send message and accept response
    '''
    websocket = yield from websockets.connect("ws://localhost:3002/")
    try:
        yield from websocket.send(message.encode())
        print("Sent: " + message)
        reply = yield from websocket.recv()
        reply = reply.decode()
    finally:
        yield from websocket.close()
    return "Got: " + reply

def build_client(server_type):
    '''
    Build a client appropriate for the server interface
    '''
    if server_type == "--websocket":
        return web_socket_client
    else:
        client = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        client.connect(("localhost", 3002))
    return client

def send_message(client, message):
    '''
    Send a message to the server and get it's response
    '''
    if isinstance(client, socket.socket):
        print("sending:", message, "Encoded:", message.encode())
        if not client.sendall(message.encode()) is None:
            print("Error in sendall of:" + message)
            print("Sent: " + message)
        print("waiting for reply")
        packet = client.recv(BUFFERSIZE).decode()
        print("Got: " + packet)
        client.close()
    else:
        print(asyncio.get_event_loop().run_until_complete(web_socket_client(message)))

def main():
    '''
    Client main
    '''
    server_type = ""
    help_string = "python3 TwitterClient.py [--websocket]\n"
    opts, args = getopt.getopt(sys.argv[1:], "", ["websocket"])
    for option in opts:
        if option[0] == "--websocket":
            server_type = option[0]
        else:
            print("Option error: " + option[0] + ", " + option[1])
            print(help_string)
            sys.exit(-1)
    send_message(build_client(server_type), " ")
    send_message(build_client(server_type), "RealSexyCyborg")
    send_message(build_client(server_type), "getEmbeddedStatus 1293170215153307660")

if __name__ == "__main__":
    main()


twitterServer.py
'''
Twitter Server - accepts websock requests, retrieves information from API
'''
import asyncio
import getopt
import json
import re
import socketserver
import sys
import websockets
from SFWIT import SFWIT

TWITTER = None

def getEmbeddedStatus(packet):
    returnInfo = TWITTER.getEmbeddedStatus(packet.split(" ")[1])
    return returnInfo

def getFollowing(packet):
    global TWITTER
    if TWITTER is None:
        TWITTER = SFWIT()
    returnInfo = TWITTER.getFollowing()
    following = json.loads(returnInfo)
    for screenName in following:
        COMMAND_LOOKUP[screenName] = getUserTimeline
    return returnInfo

def getUserTimeline(packet):
    global TWITTER
    returnInfo = TWITTER.getUserTimeline(packet)
    return returnInfo

COMMAND_LOOKUP = {
    "" : getFollowing,
    "getEmbeddedStatus" : getEmbeddedStatus}

BUFFERSIZE = 1024

def parse_command(packet):
    '''
    parse the incoming command - if null
    '''
    reply = ""
    command = packet.split(" ")[0]
    print("command:", command)
    if command in COMMAND_LOOKUP:
        print("Doing", command)
        reply = COMMAND_LOOKUP[command](packet)
    else:
        print("error:", COMMAND_LOOKUP.keys())

    if reply == "":
        return "Nak: illegal command - " + packet

    return reply

class Handler(socketserver.BaseRequestHandler):
    '''
    Handler Class for Twitter server requests
    '''

    def handle(self):
        '''
        Accept and process request
        '''
        print("packet received from client")
        packet = self.request.recv(BUFFERSIZE).strip().decode()
        print("got packet", packet)
        reply = parse_command(packet)
        print("got reply", reply)
        self.request.sendall(bytes(reply, "utf-8"))

    def finish(self):
        '''
        finish process request
        '''
        print("connection closing")


class TwitterServer(object):
    '''
    Twitter  server class
    '''
    @asyncio.coroutine
    def websocket_handler(self, websocket, path):
        '''
        Handler when running in websocket mode
        '''
        packet = yield from websocket.recv()
        packet = packet.strip().decode()
        reply = parse_command(packet)
        yield from websocket.send(bytes(reply, "utf-8"))

    def __init__(self, host, port, server_type):
        '''
        Constructor for NMI class
        '''
        if server_type == "--websocket":
            print("Starting websocket server.....")
            start_server = websockets.serve(self.websocket_handler, host, port)
            asyncio.get_event_loop().run_until_complete(start_server)
            asyncio.get_event_loop().run_forever()
        else:
            print("Starting TCP server.....")
            socketserver.TCPServer((host, port), Handler).serve_forever()

def main():
    '''
    Server main
    '''
    help_string = "python3 twitterServer.py [--websocket]"
    server_type = "" #default to TCP server
    opts, args = getopt.getopt(sys.argv[1:], "", ["websocket"])
    for option in opts:
        if option[0] == "--websocket":
            server_type = option[0]
        else:
            print("Option error: " + option[0] + ", " + option[1])
            print(help_string)
            sys.exit(-1)
    try:
        TwitterServer("localhost", 3002, server_type)
    except KeyboardInterrupt:
        print("Exiting Twitter Server gracefully")

if __name__ == "__main__":
    main()


twitter-display-server.js
// HSM display server - displays logs collected by the Host Spot Monitor
// Setup all required packages
'use strict';
var express = require("express");
var bodyParser = require("body-parser");
var fs = require("fs");
var jsdom = require("jsdom");
var readLine = require("readline");
var WebSocketServer = require("ws").Server;
var WebSocket = require("ws")
var https = require("https");
var spawn = require("child_process").spawn;
var execSync = require("child_process").execSync;
// Read main html page - this will be parsed later
let mainPageContents = fs.readFileSync("./index.html");
// Read timeline html page - this will be parsed later
let timelinePageContents = fs.readFileSync("./timeline.html");
// Read log entry page - this will be reissued later
// create an express server to make a static file server
var app = express();
var relay = [];
var getPage = function (location, row, cell1, cell2, statusID, last, response, dom) {
    let client = new WebSocket("ws://localhost:3002");
    let totalMessage = "";
    client.on("open", function(error) {
        if (error) {
            console.log("error on connection open", error);
        } else {
            // send request to get embedded status messages
            client.send(new Uint8Array(("getEmbeddedStatus " + statusID).split("").map(function (char) {return char.charCodeAt(0);})));
        }
    });
    client.on("message", function(message) {
        //console.log("received:", message);
        totalMessage += message;
    });
    client.on("close", function(message) {
        console.log("connection closed", message);
        try {
            let embeddedStatus = JSON.parse(totalMessage);
            if (typeof embeddedStatus.html !== 'undefined') {
                //console.log("Embedded Status:", embeddedStatus, embeddedStatus.html);
                let embed = new jsdom.JSDOM(embeddedStatus.html);
                //console.log(embed.window.document.body.innerHTML);
                cell2.appendChild(embed.window.document.body);
            }
        } catch (err) {
            console.log("Embedded Status translation from JSON failed, doing nothing to cell 2, error:", err);
        }
        row.appendChild(cell1);
        row.appendChild(cell2);
        location.appendChild(row);
        if (last) {
            response.send(dom.serialize());
        }
    });
    
};

// if the express server is contacted, look at the request and build a response or
// forward the request to the standard server behavior.
app.get("/", function(request, response, next) {
    let client = new WebSocket("ws://localhost:3002");
    let totalMessage = "";
    client.on("open", function(error) {
        if (error) {
            console.log("error on connection open", error);
        } else {
            // send request to get screen names that are followed
            client.send(new Uint8Array(("".split("").map(function (char) {return char.charCodeAt(0);}))));
        }
    });
    client.on("message", function(message) {
        //console.log("received:", message);
        totalMessage += message;
    });
    client.on("close", function(message) {
        console.log("connection closed", message);
        let followingObject = ["Error - none found"];
        try {
            followingObject = JSON.parse(totalMessage);
        } catch (err) {
            console.log("Error trying to translate 'following' from JSON:", err);
        }
        //console.log(followingObject);
        let dom = new jsdom.JSDOM(mainPageContents);
        let document = dom.window.document;
        let insertionPoint = document.querySelector("#list");
        for (let screenName of followingObject) {
            let element = document.createElement("a");
            let ref = "screenName_" + screenName;
            element.setAttribute("href", ref);
            element.innerHTML = screenName;
            let listElement = document.createElement("li");
            listElement.appendChild(element);
            insertionPoint.appendChild(listElement);
        }
        response.send(dom.serialize());
    });
  });
app.get("/screenName_*", function(request, response, next) {
    console.log("process a timeline route");
    let screenName = request.url.replace("/screenName_","");
    let client = new WebSocket("ws://localhost:3002")
    let totalMessage = "";
    client.on("open", function(error) {
        if (error) {
            console.log("error on connection open", error);
        } else {
            // send request to get timeline of screen name
            client.send(new Uint8Array((screenName.split("").map(function (char) {return char.charCodeAt(0);}))));
        }
    });
    client.on("message", function(message) {
        totalMessage += message;
    });
    client.on("close", function(message) {
        console.log("connection closed", message);
        let timelineObject = [];
        try {
            timelineObject = JSON.parse(totalMessage);
        } catch (err) {
            console.log("Error trying to translate timeline from JSON:", err);
        }
        let dom = new jsdom.JSDOM(timelinePageContents);
        let document = dom.window.document;
        let insertionPoint = document.querySelector("#screenName");
        insertionPoint.innerHTML = screenName;
        insertionPoint = document.querySelector("#timeline");
        let size = timelineObject.length;
        let elementsProcessed = 0;
        let processedAtLeastOneTweet = false;
        for (let onePost of timelineObject.reverse()) {
            let date = new Date(onePost[0]);
            let diff = new Date() - date;
            console.log("time diff", diff);
            elementsProcessed++;
            if (diff > 24*2*3600*1000) {
                continue;
            }
            processedAtLeastOneTweet = true;
            let tableRow = document.createElement("tr");
            let tableCell1 = document.createElement("td");
            let tableCell2 = document.createElement("td");
            let tweet = onePost[1];
            if (tweet.slice(0, 2).search("RT") >= 0) {
                tweet = onePost[1].split(':')[0] + " <br>" + onePost[2];
            }
            let pat = /https:\S+/g;
            let matches = tweet.match(pat);
            tableCell1.innerHTML = onePost[0].split("+")[0];
            if (matches) {
                let didThisTarget = []
                for (let target of matches) {
                    if ( ! didThisTarget.includes(target)) {
                        tweet = tweet.replace(target, "<a href=" + target + ">🔗</a>");
                        didThisTarget.push(target);
                    }
                }
            }
            tableCell2.innerHTML = tweet;
            let statusID = onePost[onePost.length - 1];
            getPage(insertionPoint, tableRow, tableCell1, tableCell2, statusID, elementsProcessed==size, response, dom);
        }
        if (! processedAtLeastOneTweet) {
            response.send(dom.serialize());
        }
    });
  });
app.get("*", function(request, response, next) {
    console.log("fell into default get");
    console.log(request.url);
    console.log(request.method);
    next();
  });
app.use(bodyParser.urlencoded({extended: true}));
app.use(express.static("./"));
var ws = new WebSocketServer({server: app.listen(process.env.PORT || 2999)});

ws.on("connection", function(connection) {
    relay.push(connection); // store for communication
    console.log("web socket connection made at server from HTML client page");
    connection.send("connected");
    connection.on("message", function (message) {
        if (message === "exit") {
          relay.splice(relay.indexOf(connection), 1);
          connection.close();
        }
      });
    connection.on("close", function(message) {
        relay.splice(relay.indexOf(connection), 1);
        connection.close();
        console.log("closing a connection");
      });
  });

console.log("Twitter display server is listening");
